### ISSUE TITLE
anyOf should not result in intersection type

### ISSUE THREAD

**Issue by MunifTanjim** [drwpow/openapi-typescript#894]
`anyOf` is equivalent to `OR`, i.e. TypeScript's union `|` operator.

The current behavior is incorrect.

---

Related Note:

`oneOf` is equivalent to `XOR`, no direct equivalent for TypeScript. Currently treated as `OR`.

**Comment by drwpow** [comment]
Hm, from the [OpenAPI specification](https://swagger.io/docs/specification/data-models/oneof-anyof-allof-not/):

> OpenAPI 3.0 provides several keywords which you can use to combine schemas. You can use these keywords to create a complex schema, or validate a value against multiple criteria.
>
>    - [`oneOf`](https://swagger.io/docs/specification/data-models/oneof-anyof-allof-not/#oneof) – validates the value against exactly one of the subschemas
>    - [`allOf`](https://swagger.io/docs/specification/data-models/oneof-anyof-allof-not/#allof) – validates the value against all the subschemas
>    - [`anyOf`](https://swagger.io/docs/specification/data-models/oneof-anyof-allof-not/#anyof) – validates the value against any (one or more) of the subschemas

I don’t believe `anyOf` is `OR`, as it may be either **one** of the subschemas, or **all** of the subschemas. To get `OR` you would want `oneOf`, no?

I think the current behavior is correct (`Partial<A> & Partial<B> & Partial<C>`) unless I’m missing something.

**Comment by MunifTanjim** [comment]
Say you want to define a parameter that accepts either `number` or `string`. If you do `number & string` or `Partial<number> & Partial<string>` in TypeScript, both of them result in `never`. Which is clearly not what we want.

Now think of this type in TypeScript: `type Str = string | 'one'`. Any string value (including `'one'`) will satisfy the `Str` type. It's not exclusively `oneOf` `string` or `'one'`, because in that case `'one'` is an impossible value (it can't be exclusively 'one', but not `string`).

From the OpenAPI's specification, `oneOf` is Exclusive OR - which can't be described easily in TypeScript without writing some complex custom types. And `anyOf` is essentially equivalent to the union operator `|`.

Does the explanation sound reasonable?

**Comment by MunifTanjim** [comment]
Here's another example.

With the current behavior:

```yaml
paths:
  /test:
    post:
      requestBody:
        content:
          application/json:
            schema:
              anyOf:
                - $ref: '#/components/schemas/A'
                - $ref: '#/components/schemas/B'
      responses:
        '200':
          description: ''
components:
  schemas:
    A:
      type: object
      properties: 
        type: 
          type: string
          enum: ['a']
        a: 
          type: number
      required:
        - type
        - a
          
    B:
      type: object
      properties:
        type: 
          type: string
          enum: ['b']
        b: 
          type: string
      required: 
        - type
        - b
```

would result in:

```ts
type A = {
  type: 'a',
  a: number
}

type B = {
  type: 'b',
  b: string
}

type X = Partial<A> & Partial<B>

// testing
const x1: X = {}
const x2: X = { a: 1 }
```

If a parameter accepts `anyOf` type `A` or type `B`, the TypeScript types generated by `openapi-typescript` is `Partial<A> & Partial<B>`, which passes for:

- `{}` - but empty object is not accepted at all.
- `{ a: 1 }` but `a: number` is only supported when `type: 'a'`.

> I think the current behavior is correct (Partial<A> & Partial<B> & Partial<C>)

So the current behavior is giving false positives.

**Comment by drwpow** [comment]
Yes I see the improvement that could be made here. That short summary I posted earlier is a bit confusing, but reading more into it, it does seem that **`anyOf` requires it to validate against at least one of the schemas** which `Partial<>` would indeed prevent, as you pointed out. I think `Partial<>` should be removed regardless.

So in digging into this a little more, I learned a little more about TypeScript union types in relation to objects, and they behave a little differently than expected. For example:

```ts
type AnyOf = string | { a: string } | { b: string };

const testA: AnyOf = 'foo';                 // ✅ valid
const testB: AnyOf = { a: 'foo' }           // ✅ valid
const testC: AnyOf = { b: 'foo' }           // ✅ valid
const testD: AnyOf = { a: 'foo', b: 'foo' } // ✅ valid
```

It seems that TypeScript unions provide proper `anyOf` behavior on their own, without modification. So I think just a simple union type would suffice for `anyOf`?

However, this does pose a problem for `oneOf`:

```ts
type OneOf = string | { a: string } | { b: string };

const testA: OneOf = 'foo';                 // ✅ valid
const testB: OneOf = { a: 'foo' }           // ✅ valid
const testC: OneOf = { b: 'foo' }           // ✅ valid
const testD: OneOf = { a: 'foo', b: 'foo' } // ❌ this is valid but shouldn’t be
```

I’m not sure how to solve this problem; [one solution I found](https://stackoverflow.com/questions/44425344/typescript-interface-with-xor-barstring-xor-cannumber) was to provide `never` for all keys, like so:

```
type OneOf = string | { a: string; b: never } | { a: never; b: string };'

const testD: OneOf = { a: 'foo', b: 'foo' } // ✅ this now fails the typecheck
```

It seems to me like `anyOf` should change to just a simple **union** type, like `oneOf` currently implements. However, `oneOf`’s behavior should also change to not allow intersections to happen like they do currently. Whether that’s the `never` approach or a better one, I’m not sure. And I’m tempted to make the latter a separate scope of work entirely (especially if it’s breaking behavior).

Does this proposal sound reasonable?
